File: src/main/java/yaasl/server/convert/Converter.java
package yaasl.server.convert;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import yaasl.server.jsonapi.Element;
import yaasl.server.jsonapi.SingleData;
import yaasl.server.model.*;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.format.DateTimeFormatter;
import java.util.*;

import static java.lang.Boolean.parseBoolean;
import static java.lang.Long.parseLong;
import static java.time.format.DateTimeFormatter.ISO_DATE_TIME;
import static org.apache.commons.lang3.StringUtils.isNotEmpty;
import static org.apache.commons.lang3.time.DateUtils.truncate;

public class Converter {

    private static DateTimeFormatter isoDateTimeFormat = ISO_DATE_TIME;
    private static SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
    private static SimpleDateFormat dateTimeFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");

    private static Logger LOG = LoggerFactory.getLogger(Converter.class);

    public static Element convert(Location location) {
        return convert(location, true);
    }

    public static Element convert(Location location, boolean includeAttributes) {
        Element element = new Element();
        element.setId(location.getId().toString());
        element.setType("location");
        if (includeAttributes) {
            element.addAttribute("name", location.getName());
        }
        return element;
    }

    public static Element convert(Aircraft aircraft) {
        return convert(aircraft, true);
    }

    public static Element convert(Aircraft aircraft, boolean includeAttributes) {
        Element element = new Element();
        element.setId(aircraft.getId().toString());
        element.setType("aircraft");
        if (includeAttributes) {
            element.addAttribute("call-sign", aircraft.getCallSign());
            element.addAttribute("can-tow", aircraft.isCanTow());
            element.addAttribute("needs-towing", aircraft.isNeedsTowing());
            element.addAttribute("number-of-seats", aircraft.getNumberOfSeats());
        }
        return element;
    }

    public static Element convert(Pilot pilot) {
        return convert(pilot, true);
    }

    public static Element convert(Pilot pilot, boolean includeAttributes) {
        Element element = new Element();
        element.setId(pilot.getId().toString());
        element.setType("pilot");
        if (includeAttributes) {
            element.addAttribute("name", pilot.getName());
            element.addAttribute("can-tow", pilot.isCanTow());
        }
        if (pilot.getStandardRole() != null) {
            element.addRelationship("standard-role", new SingleData(convert(pilot.getStandardRole(), false)));
        }
        return element;
    }

    public static Element convert(PilotRole pilotRole) {
        return convert(pilotRole, true);
    }

    public static Element convert(PilotRole pilotRole, boolean includeAttributes) {
        Element element = new Element();
        element.setId(pilotRole.getId().toString());
        element.setType("pilot-role");
        if (includeAttributes) {
            element.addAttribute("description", pilotRole.getDescription());
            element.addAttribute("i18n", pilotRole.getI18n());
        }
        return element;
    }

    public static Element convert(CostSharing costSharing) {
        return convert(costSharing, true);
    }

    public static Element convert(CostSharing costSharing, boolean includeAttributes) {
        Element element = new Element();
        element.setId(costSharing.getId().toString());
        element.setType("cost-sharing");
        if (includeAttributes) {
            element.addAttribute("description", costSharing.getDescription());
            element.addAttribute("i18n", costSharing.getI18n());
        }
        return element;
    }

    public static Element convert(Flight flight) {
        Element element = new Element();
        element.setId(flight.getId().toString());
        element.setType("flight");
        if (flight.getStartTime() != null) {
            element.addAttribute("start-time", dateTimeFormat.format(flight.getStartTime()));
        }
        if (flight.getStartLocation() != null) {
            element.addRelationship("start-location", new SingleData(convert(flight.getStartLocation(), false)));
        }
        if (flight.getLandingTime() != null) {
            element.addAttribute("landing-time", dateTimeFormat.format(flight.getLandingTime()));
        }
        if (flight.getLandingLocation() != null) {
            element.addRelationship("landing-location", new SingleData(convert(flight.getLandingLocation(), false)));
        }
        if (flight.getAircraft() != null) {
            element.addRelationship("aircraft", new SingleData(convert(flight.getAircraft(), false)));
        }
        if (flight.getPilot1() != null) {
            element.addRelationship("pilot1", new SingleData(convert(flight.getPilot1(), false)));
        }
        if (flight.getPilot1Role() != null) {
            element.addRelationship("pilot1-role", new SingleData(convert(flight.getPilot1Role(), false)));
        }
        if (flight.getPilot2() != null) {
            element.addRelationship("pilot2", new SingleData(convert(flight.getPilot2(), false)));
        }
        if (flight.getPilot2Role() != null) {
            element.addRelationship("pilot2-role", new SingleData(convert(flight.getPilot2Role(), false)));
        }
        if (flight.getTowPlane() != null) {
            element.addRelationship("tow-plane", new SingleData(convert(flight.getTowPlane(), false)));
        }
        if (flight.getTowPilot() != null) {
            element.addRelationship("tow-pilot", new SingleData(convert(flight.getTowPilot(), false)));
        }
        if (flight.getTowPlaneLandingTime() != null) {
            element.addAttribute("tow-plane-landing-time", dateTimeFormat.format(flight.getTowPlaneLandingTime()));
        }
        if (flight.getCostSharing() != null) {
            element.addRelationship("cost-sharing", new SingleData(convert(flight.getCostSharing(), false)));
        }
        if (isNotEmpty(flight.getComment())) {
            element.addAttribute("comment", flight.getComment());
        }
        element.addAttribute("locked", flight.isLocked());
        return element;
    }

    public static Flight convert(Element element) throws Exception {
        Map<String, Object> attributes = element.getAttributes();
        Flight flight = new Flight();
        if (element.getId() != null) {
            flight.setId(parseLong(element.getId()));
        }
        flight.setStartLocation(convertLocation(getRelationship("start-location", element)));
        flight.setStartTime(parseDateTime((String) attributes.get("start-time")));
        flight.setLandingLocation(convertLocation(getRelationship("landing-location", element)));
        flight.setLandingTime(parseDateTime(((String) attributes.get("landing-time"))));
        flight.setAircraft(convertAircraft(getRelationship("aircraft", element)));
        flight.setPilot1(convertPilot(getRelationship("pilot1", element)));
        flight.setPilot1Role(convertPilotRole(getRelationship("pilot1-role", element)));
        flight.setPilot2(convertPilot(getRelationship("pilot2", element)));
        flight.setPilot2Role(convertPilotRole(getRelationship("pilot2-role", element)));
        flight.setTowPlane(convertAircraft(getRelationship("tow-plane", element)));
        flight.setTowPilot(convertPilot(getRelationship("tow-pilot", element)));
        flight.setTowPlaneLandingTime(parseDateTime(((String) attributes.get("tow-plane-landing-time"))));
        flight.setCostSharing(convertCostSharing(getRelationship("cost-sharing", element)));
        flight.setComment((String) attributes.get("comment"));
        flight.setLocked(attributes.get("locked") != null ? (boolean) attributes.get("locked") : true);
        return flight;
    }

    public static Date parseDate(String date) {
        try {
            return dateFormat.parse(date);
        } catch (ParseException e) {
            return null;
        }
    }

    public static String formatDate(Date date) {
        return dateFormat.format(date);
    }

    private static Location convertLocation(Map<String, Object> json) throws IOException {
        if (json != null) {
            Location location = new Location();
            location.setId(parseLong((String) ((Map<String, Object>) json.get("data")).get("id")));
            return location;
        } else {
            return null;
        }
    }

    private static Aircraft convertAircraft(Map<String, Object> json) throws IOException {
        if (json != null) {
            Aircraft aircraft = new Aircraft();
            aircraft.setId(parseLong((String) ((Map<String, Object>) json.get("data")).get("id")));
            return aircraft;
        } else {
            return null;
        }
    }

    private static Pilot convertPilot(Map<String, Object> json) throws IOException {
        if (json != null) {
            Pilot pilot = new Pilot();
            pilot.setId(parseLong((String) ((Map<String, Object>) json.get("data")).get("id")));
            return pilot;
        } else {
            return null;
        }
    }

    private static PilotRole convertPilotRole(Map<String, Object> json) throws IOException {
        if (json != null) {
            PilotRole pilotRole = new PilotRole();
            pilotRole.setId(parseLong((String) ((Map<String, Object>) json.get("data")).get("id")));
            return pilotRole;
        } else {
            return null;
        }
    }

    private static CostSharing convertCostSharing(Map<String, Object> json) throws IOException {
        if (json != null) {
            CostSharing costSharing = new CostSharing();
            costSharing.setId(parseLong((String) ((Map<String, Object>) json.get("data")).get("id")));
            return costSharing;
        } else {
            return null;
        }
    }

    private static Map<String, Object> getRelationship(String name, Element element) {
        if (element.getRelationships() != null) {
            for (Map.Entry<String, Object> relationship : element.getRelationships().entrySet()) {
                if (name.equals(relationship.getKey())) {
                    return (Map<String, Object>) relationship.getValue();
                }
            }
        }
        return null;
    }

    private static Date parseDateTime(String dateTime) {
        Date date = null;
        if (isNotEmpty(dateTime)) {
            try {
                date = dateTimeFormat.parse(dateTime);
            } catch (ParseException e) {
                LOG.error("Unable to convert {}", dateTime, e);
            }
        }
        return date;
    }

}

File: src/main/kotlin/yaasl/server/convert/Converter.kt
package yaasl.server.convert

import org.apache.commons.lang3.StringUtils.isNotEmpty
import org.slf4j.LoggerFactory
import yaasl.server.jsonapi.Element
import yaasl.server.jsonapi.SingleData
import yaasl.server.model.*
import java.lang.Long.parseLong
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

object Converter {

    private val dateFormat = SimpleDateFormat("yyyy-MM-dd")
    private val dateTimeFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")

    private val LOG = LoggerFactory.getLogger(Converter::class.java)

    fun convert(location: Location, includeAttributes: Boolean = true): Element {
        val element = Element()
        element.setId(location.id.toString())
        element.setType("location")
        if (includeAttributes) {
            element.addAttribute("icao", location.icao!!)
            element.addAttribute("name", location.name!!)
        }
        return element
    }

    fun convert(aircraft: Aircraft, includeAttributes: Boolean = true): Element {
        val element = Element()
        element.setId(aircraft.id.toString())
        element.setType("aircraft")
        if (includeAttributes) {
            element.addAttribute("call-sign", aircraft.callSign)
            element.addAttribute("competition-number", aircraft.competitionNumber)
            element.addAttribute("can-tow", aircraft.canTow)
            element.addAttribute("needs-towing", aircraft.needsTowing)
            element.addAttribute("number-of-seats", aircraft.numberOfSeats)
        }
        return element
    }


    fun convert(pilot: Pilot, includeAttributes: Boolean = true): Element {
        val element = Element()
        element.setId(pilot.id.toString())
        element.setType("pilot")
        if (includeAttributes) {
            element.addAttribute("name", pilot.name!!)
            element.addAttribute("can-tow", pilot.isCanTow)
        }
        if (pilot.standardRole != null) {
            element.addRelationship("standard-role", SingleData(convert(pilot.standardRole!!, false)))
        }
        return element
    }

    fun convert(pilotRole: PilotRole, includeAttributes: Boolean = true): Element {
        val element = Element()
        element.setId(pilotRole.id.toString())
        element.setType("pilot-role")
        if (includeAttributes) {
            element.addAttribute("description", pilotRole.description!!)
            element.addAttribute("i18n", pilotRole.i18n!!)
        }
        return element
    }

    fun convert(costSharing: CostSharing, includeAttributes: Boolean = true): Element {
        val element = Element()
        element.setId(costSharing.id.toString())
        element.setType("cost-sharing")
        if (includeAttributes) {
            element.addAttribute("description", costSharing.description!!)
            element.addAttribute("i18n", costSharing.i18n!!)
        }
        return element
    }

    fun convert(flight: Flight): Element {
        val element = Element()
        element.setId(flight.id.toString())
        element.setType("flight")
        if (flight.startTime != null) {
            element.addAttribute("start-time", dateTimeFormat.format(flight.startTime))
        }
        if (flight.startLocation != null) {
            element.addRelationship("start-location", SingleData(convert(flight.startLocation!!, false)))
        }
        if (flight.landingTime != null) {
            element.addAttribute("landing-time", dateTimeFormat.format(flight.landingTime))
        }
        if (flight.landingLocation != null) {
            element.addRelationship("landing-location", SingleData(convert(flight.landingLocation!!, false)))
        }
        if (flight.aircraft != null) {
            element.addRelationship("aircraft", SingleData(convert(flight.aircraft!!, false)))
        }
        if (flight.pilot1 != null) {
            element.addRelationship("pilot1", SingleData(convert(flight.pilot1!!, false)))
        }
        if (flight.pilot1Role != null) {
            element.addRelationship("pilot1-role", SingleData(convert(flight.pilot1Role!!, false)))
        }
        if (flight.pilot2 != null) {
            element.addRelationship("pilot2", SingleData(convert(flight.pilot2!!, false)))
        }
        if (flight.pilot2Role != null) {
            element.addRelationship("pilot2-role", SingleData(convert(flight.pilot2Role!!, false)))
        }
        if (flight.towPlane != null) {
            element.addRelationship("tow-plane", SingleData(convert(flight.towPlane!!, false)))
        }
        if (flight.towPilot != null) {
            element.addRelationship("tow-pilot", SingleData(convert(flight.towPilot!!, false)))
        }
        if (flight.towPlaneLandingTime != null) {
            element.addAttribute("tow-plane-landing-time", dateTimeFormat.format(flight.towPlaneLandingTime))
        }
        if (flight.costSharing != null) {
            element.addRelationship("cost-sharing", SingleData(convert(flight.costSharing!!, false)))
        }
        if (isNotEmpty(flight.comment)) {
            element.addAttribute("comment", flight.comment!!)
        }
        element.addAttribute("editable", flight.editable)
        element.addAttribute("locked", flight.locked)
        element.addAttribute("revision", flight.revision)
        return element
    }

    fun convert(feedback: Feedback): Element {
        val element = Element()
        element.setId(feedback.id.toString())
        element.setType("feedback")
        if (feedback.feedback != null) {
            element.addAttribute("feedback", feedback.feedback!!)
        }
        if (feedback.comment != null) {
            element.addAttribute("comment", feedback.comment!!)
        }
        return element
    }

    fun <T> convert(element: Element): T? {
        val attributes = element.getAttributes()
        when (element.getType()) {
            "flights" -> {
                val flight = Flight()
                if (element.getId() != null) {
                    flight.id = parseLong(element.getId())
                }
                flight.startLocation = convertLocation(getRelationship("start-location", element))
                flight.startTime = parseDateTime(attributes["start-time"] as? String)
                flight.landingLocation = convertLocation(getRelationship("landing-location", element))
                flight.landingTime = parseDateTime(attributes["landing-time"] as? String)
                flight.aircraft = convertAircraft(getRelationship("aircraft", element))
                flight.pilot1 = convertPilot(getRelationship("pilot1", element))
                flight.pilot1Role = convertPilotRole(getRelationship("pilot1-role", element))
                flight.pilot2 = convertPilot(getRelationship("pilot2", element))
                flight.pilot2Role = convertPilotRole(getRelationship("pilot2-role", element))
                flight.towPlane = convertAircraft(getRelationship("tow-plane", element))
                flight.towPilot = convertPilot(getRelationship("tow-pilot", element))
                flight.towPlaneLandingTime = parseDateTime(attributes["tow-plane-landing-time"] as? String)
                flight.costSharing = convertCostSharing(getRelationship("cost-sharing", element))
                flight.comment = attributes["comment"] as? String
                flight.editable = getAttribute("editable", false, attributes)
                flight.locked = getAttribute("locked", true, attributes)
                flight.revision = getAttribute("revision", 0, attributes)?.toLong()
                return flight as T
            }
            "feedbacks" -> {
                val feedback = Feedback()
                feedback.feedback = getAttribute<String>("feedback", null, attributes)
                feedback.comment = getAttribute<String>("comment", null, attributes)
                return feedback as T
            }
        }
        return null
    }

    fun parseDate(date: String): Date? {
        try {
            return dateFormat.parse(date)
        } catch (e: ParseException) {
            return null
        }
    }

    fun formatDate(date: Date): String {
        return dateFormat.format(date)
    }

    private fun convertLocation(json: Map<String, Any>?): Location? {
        if (json != null) {
            val location = Location()
            location.id = parseLong((json["data"] as Map<String, Any>)["id"] as String)
            return location
        } else {
            return null
        }
    }

    private fun convertAircraft(json: Map<String, Any>?): Aircraft? {
        if (json != null) {
            val aircraft = Aircraft()
            aircraft.id = parseLong((json["data"] as Map<String, Any>)["id"] as String)
            return aircraft
        } else {
            return null
        }
    }

    private fun convertPilot(json: Map<String, Any>?): Pilot? {
        if (json != null) {
            val pilot = Pilot()
            pilot.id = parseLong((json["data"] as Map<String, Any>)["id"] as String)
            return pilot
        } else {
            return null
        }
    }

    private fun convertPilotRole(json: Map<String, Any>?): PilotRole? {
        if (json != null) {
            val pilotRole = PilotRole()
            pilotRole.id = parseLong((json["data"] as Map<String, Any>)["id"] as String)
            return pilotRole
        } else {
            return null
        }
    }

    private fun convertCostSharing(json: Map<String, Any>?): CostSharing? {
        if (json != null) {
            val costSharing = CostSharing()
            costSharing.id = parseLong((json["data"] as Map<String, Any>)["id"] as String)
            return costSharing
        } else {
            return null
        }
    }

    private fun getRelationship(name: String, element: Element): Map<String, Any>? {
        for ((key, value) in element.getRelationships() as Map<String, Any>) {
            if (name == key) {
                return value as Map<String, Any>
            }
        }
        return null
    }

    private fun parseDateTime(dateTime: String?): Date? {
        var date: Date? = null
        if (dateTime != null && dateTime.isNotEmpty()) {
            try {
                date = dateTimeFormat.parse(dateTime)
            } catch (e: ParseException) {
                LOG.error("Unable to convert {}", dateTime, e)
            }

        }
        return date
    }

    fun <T> getAttribute(attributeName: String, defaultValue: T?, attributes: Map<String, Any?>): T? {
        return if (attributes[attributeName] != null) {
            attributes[attributeName] as T
        } else {
            defaultValue
        }
    }

}
